fixes for TS types and added types to unit tests

xaviergonz opened the pull request:
basically a (kinda big) pull request to fix lots of TS typing issues

btw, shouldn't custom type actually use <C, S, T> rather than <S|T,S,T> for more flexibility?

xaviergonz commented:
There's one unit test failing:
- it should emit snapshots for children
but it also fails in master
```
 it should emit snapshots for children

    expect(received).toEqual(expected)

    Expected value to equal:
      {"files": [{"name": "01-arrival"}, {"name": "Photo2"}], "name": "Vacation photos"}
    Received:
      {"name": "01-arrival"}

    Difference:

    - Expected
    + Received

      Object {
    -   "files": Array [
    -     Object {
        "name": "01-arrival",
    -     },
    -     Object {
    -       "name": "Photo2",
    -     },
    -   ],
    -   "name": "Vacation photos",
      }

      139 |     onSnapshot(folder.files[0], snapshot => snapshots.push(snapshot))
      140 |     folder.files[0].rename("01-arrival")
    > 141 |     expect(snapshots[1]).toEqual({
          |                          ^
      142 |         name: "Vacation photos",
      143 |         files: [{ name: "01-arrival" }, { name: "Photo2" }]
      144 |     })
```

xaviergonz commented:
Apparently the error was an accidental change to mobx-state-tree package-json that changed mobx version to 4.3.1 from 5.0.3.
I restored it and now everything seems to be fine.

benjaminjackman commented on packages/mobx-state-tree/package.json:
```diff
@@ -41,7 +41,7 @@
     "documentation": "^5.2.2",
     "jest": "^23.2.0",
-    "mobx": "5.0.3",
+    "mobx": "4.3.1",
```
is that intentional?

xaviergonz replied:
accidental, already reverted

xaviergonz commented:
besides making TypeFlags public and exported also added TypeFlags imports to undo-manager and the other mst-middleware to fix https://github.com/mobxjs/mobx-state-tree/issues/930

xaviergonz commented:
Also fixes https://github.com/mobxjs/mobx-state-tree/issues/932

xaviergonz commented:
just added a fix (with unit test) for optional enums inside a model crashing
in other words, the following will crash:
```
test("should support optional enums inside a model", () => {
    const TrafficLight = types.model({
        color: types.optional(types.enumeration(["Red", "Orange", "Green"]), "Orange")
    })
    const l = TrafficLight.create({})
    expect(l.color).toBe("Orange")
})
```

because types.optional would inherit as flag TypeFlags.Union, but not "types" and therefore crash here
```
function tryGetOptional(type: any): OptionalValue<any, any, any> | undefined {
    if (!type) return undefined
    if (type.flags & TypeFlags.Union) return type.types.find(tryGetOptional) // THIS LINE
    if (type.flags & TypeFlags.Late) return tryGetOptional(type.subType)
    if (type.flags & TypeFlags.Optional) return type
    return undefined
}
```

so I fixed it like this
```
function tryGetOptional(type: any): OptionalValue<any, any, any> | undefined {
    if (!type) return undefined
    // we need to check for type.types since an optional union doesn't have direct subtypes
    if (type.flags & TypeFlags.Union && type.types) return type.types.find(tryGetOptional)
    if (type.flags & TypeFlags.Late && type.subType) return tryGetOptional(type.subType)
    if (type.flags & TypeFlags.Optional) return type
    return undefined
}
```

another option would be to make optional just return the optional flag, but I'm not sure if that's better

mweststrate commented:
Thanks for the investigation! I think the solution is correct, will double
check next week!

Op vr 20 jul. 2018 om 20:01 schreef Javier Gonzalez <
notifications@github.com>:

> just added a fix (with unit test) for optional enums inside a model
> crashing
> in other words, the following will crash:
>
> test("should support optional enums inside a model", () => {
>     const TrafficLight = types.model({
>         color: types.optional(types.enumeration(["Red", "Orange", "Green"]), "Orange")
>     })
>     const l = TrafficLight.create({})
>     expect(l.color).toBe("Orange")
> })
>
> because types.optional would inherit as flag TypeFlags.Union, but not
> "types" and therefore crash here
>
> function tryGetOptional(type: any): OptionalValue<any, any, any> | undefined {
>     if (!type) return undefined
>     if (type.flags & TypeFlags.Union) return type.types.find(tryGetOptional) // THIS LINE
>     if (type.flags & TypeFlags.Late) return tryGetOptional(type.subType)
>     if (type.flags & TypeFlags.Optional) return type
>     return undefined
> }
>
> so I fixed it like this
>
> function tryGetOptional(type: any): OptionalValue<any, any, any> | undefined {
>     if (!type) return undefined
>     // we need to check for type.types since an optional union doesn't have direct subtypes
>     if (type.flags & TypeFlags.Union && type.types) return type.types.find(tryGetOptional)
>     if (type.flags & TypeFlags.Late && type.subType) return tryGetOptional(type.subType)
>     if (type.flags & TypeFlags.Optional) return type
>     return undefined
> }
>
> another option would be to make optional just return the optional flag,
> but I'm not sure if that's better
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-406680944>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvGhG2ySiJCLS3LoSoN1meNrL0Wjogaks5uIhsVgaJpZM4VXR1D>
> .
>

xaviergonz commented:
Maybe fixes #922

xaviergonz commented:
Ok, basically I took the opportunity to make a revamp of the typescript typing to further improve it.
Notable changes is that ISnapshottable is no longer there and its role has been taken by IStateTreeNode<S>

With these changes now getSnapshot is properly typed without the need to use its generic argument, references and identifiers work properly and lots of other little fixes.

xaviergonz commented:
I made this available in npm as ```@xaviergonz/mobx-state-tree @xaviergonz/mst-middlewares``` for anybody that might want to give it a try btw

xaviergonz commented:
I had to remove DeepImmutable. While cool (made snapshots from getSnapshot unmodifiable, etc) I have a large project that uses MST and it was getting out of ram errors while compiling because of it.

That being said, I'm quite happy with the current state of the typings 👍 so I don't think I'll make further changes.

xaviergonz commented:
@mweststrate my pleasure :)

mweststrate commented:
Took a close look and these are great improvements! Really awesome work. I've just send you a maintainers invite, that will simplify the workflow for any future contributions a lot, as you can directly push, branch etc on this repo :-)!

dpnolte commented:
@xaviergonz  thanks this fixes my issue 👍 
I hope this get published soon on npm.

marvwhere commented:
cool thx - tried it out (v3.0.2) 

looks like it's causing some problems with fork-ts-checker-webpack-plugin - "Starting type checking and linting service..." will never finish and creating a node instance with >100% cpu

went back to 2.2.0 (working)
updated to 3.0.0 (working)
updated to 3.0.1 (not working)

Typescript 2.9
tslint 5.11.0

mweststrate commented:
and if you disable tslint?


Op do 26 jul. 2018 om 16:04 schreef Marvin <notifications@github.com>:

> cool thx - tried it out (v3.0.2)
>
> looks like it's causing some problems with fork-ts-checker-webpack-plugin
> - "Starting type checking and linting service..." will never finish and
> creating a node instance with >100% cpu
>
> went back to 2.2.0 (working)
> updated to 3.0.0 (working)
> updated to 3.0.1 (not working)
>
> Typescript 2.9
> tslint 5.11.0
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-408108712>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvGhPzuhFVPpsRD7snAf5q4Y4rEJgT3ks5uKcxxgaJpZM4VXR1D>
> .
>

xaviergonz commented:
I have seen that happen from time to time when there are lots of compilation errors. Basically the default node memory of node processes is around 512mb and complex type checking takes lots of ram and then crash.

Things I'd try include:
- run tsc by itself on the command line: ```npx tsc -p .```
- give it more ram if it dies because of out of ram: ```NODE_OPTIONS=--max-old-space-size=4096 && tsc -p .```
- make use of the speed up trick (interface instead of typeof)
```
type CreationType = typeof SomeModel.CreationType
type SnapshotType = typeof SomeModel.SnapshotType
type Type = typeof SomeModel.Type
export interface ISomeModelCreation extends CreationType {}
export interface ISomeModelSnapshot extends SnapshotType {}
export interface ISomeModel extends Type {}
```
then using ISomeModelCreation, ISomeModelSnapshot, ISomeModel 

or I guess TS could get their act together and optimize that :)

Bnaya commented:
```ts
import { types, getSnapshot } from "mobx-state-tree";

const Square = types
    .model(
        "Square",
        {
            width: types.number
        }
    )

const a = getSnapshot(Square)
```
I'm getting this error for the getSnapshot call
`
Type 'IModelType<ModelPropertiesDeclarationToProperties<{ width: ISimpleType<number>; }>, {}, ModelCrea...' has no properties in common with type 'IStateTreeNode<any, {}>'.`

I'm probably missing something :)

mweststrate commented:
Yes, you are snapshoitting a type, not an instance

Op do 26 jul. 2018 20:38 schreef Bnaya Peretz <notifications@github.com>:

> import { types, getSnapshot } from "mobx-state-tree";
> const Square = types
>     .model(
>         "Square",
>         {
>             width: types.number
>         }
>     )
> const a = getSnapshot(Square)
>
> I'm getting this error for the getSnapshot call
> Type 'IModelType<ModelPropertiesDeclarationToProperties<{ width:
> ISimpleType<number>; }>, {}, ModelCrea...' has no properties in common with
> type 'IStateTreeNode<any, {}>'.
>
> I'm probably missing something :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-408194750>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvGhIT0NtIxj1xFFgDcdTxrINOzVkhdks5uKgzBgaJpZM4VXR1D>
> .
>

Bnaya commented:
Of Course😂 thanks

On Thu, Jul 26, 2018, 21:43 Michel Weststrate <notifications@github.com>
wrote:

> Yes, you are snapshoitting a type, not an instance
>
> Op do 26 jul. 2018 20:38 schreef Bnaya Peretz <notifications@github.com>:
>
> > import { types, getSnapshot } from "mobx-state-tree";
> > const Square = types
> > .model(
> > "Square",
> > {
> > width: types.number
> > }
> > )
> > const a = getSnapshot(Square)
> >
> > I'm getting this error for the getSnapshot call
> > Type 'IModelType<ModelPropertiesDeclarationToProperties<{ width:
> > ISimpleType<number>; }>, {}, ModelCrea...' has no properties in common
> with
> > type 'IStateTreeNode<any, {}>'.
> >
> > I'm probably missing something :)
> >
> > —
> > You are receiving this because you were mentioned.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-408194750
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/ABvGhIT0NtIxj1xFFgDcdTxrINOzVkhdks5uKgzBgaJpZM4VXR1D
> >
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-408196032>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPpHuNjGJhGVxQ-6E5G6EqUhyb_Fadxks5uKg3BgaJpZM4VXR1D>
> .
>

marvwhere commented:
@xaviergonz I tried around, but still not helping. The strange thing is that it is working with 3.0.0. I used that version to remove all lint errors which popped up through the breaking changes. then went back to 3.0.2 - still not working (i mean it is compiling, but the linter is not working)

I try to play around with it a little bit and try to find out which commit breaks it.

mweststrate commented:
Sounds more like an issue to report at the linter library?

Op vr 27 jul. 2018 11:26 schreef Javier Gonzalez <notifications@github.com>:

> so the compiler works but the linter doesn't? what if you just run tslint
> stand alone?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mobxjs/mobx-state-tree/pull/937#issuecomment-408364751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvGhJXOdJN90I1WB0qPKfD_8awc_5v3ks5uKtzdgaJpZM4VXR1D>
> .
>

marvwhere commented:
kind of yes =/ 

`NODE_OPTIONS=--max-old-space-size=4096 && npx tsc -p .` command 

working fine in 3.0.0 and does nothing in 3.0.1
